autoload/lsp/ui/vim/semantic.vim

```diff
@@ -0,0 +1,194 @@
+        return []
+    endif
+
+    let l:capabilities = lsp#get_server_capabilities(a:server)
+    return l:capabilities['semanticHighlighting']['scopes']
+endfunction
+
+function! lsp#ui#vim#semantic#handle_semantic(server, data) abort
+    if !g:lsp_semantic_enabled | return | endif
+
+    if lsp#client#is_error(a:data['response'])
+        call lsp#log('Skipping semantic highlight: response is invalid')
+        return
+    endif
+
+    let l:uri = a:data['response']['params']['textDocument']['uri']
+    let l:path = lsp#utils#uri_to_path(l:uri)
+    let l:bufnr = bufnr(l:path)
+
+    if l:bufnr < 0 | return | endif
+
+    " Skip if the buffer doesn't exist. This might happen when a buffer is
+    " opened and quickly deleted.
+    if !bufloaded(l:bufnr) | return | endif
+
+    call s:init_highlight(a:server, l:bufnr)
+
+    for l:info in a:data['response']['params']['lines']
+        let l:linenr = l:info['line']
+        let l:tokens = has_key(l:info, 'tokens') ? l:info['tokens'] : ''
+        call s:add_highlight(a:server, l:bufnr, l:linenr, s:tokens_to_hl_info(l:tokens))
+    endfor
+endfunction
+
+" Highlight helper functions {{{1
+function! s:init_highlight(server, buf) abort
+    if !empty(getbufvar(a:buf, 'lsp_did_semantic_setup'))
+        return
+    endif
+
+    if s:use_vim_textprops
+        let l:scopes = lsp#ui#vim#semantic#get_scopes(a:server)
+        for l:scope_idx in range(len(l:scopes))
+            let l:scope = l:scopes[l:scope_idx]
+            let l:hl = s:get_hl_name(a:server, l:scope)
+
+            silent! call prop_type_add(s:get_textprop_name(a:server, l:scope_idx), {'bufnr': a:buf, 'highlight': l:hl, 'combine': v:true})
+        endfor
+    endif
+
+    call setbufvar(a:buf, 'lsp_did_semantic_setup', 1)
+endfunction
+
+function! s:add_highlight(server, buf, line, highlights) abort
+    let l:scopes = lsp#ui#vim#semantic#get_scopes(a:server)
+
+    if s:use_vim_textprops
+        " Clear text properties from the previous run
+        for l:scope_idx in range(len(l:scopes))
+            call prop_remove({'bufnr': a:buf, 'type': s:get_textprop_name(a:server, l:scope_idx), 'all': v:true}, a:line + 1)
+        endfor
+
+        for l:highlight in a:highlights
+            call prop_add(a:line + 1, l:highlight['char'] + 1, { 'length': l:highlight['length'], 'bufnr': a:buf, 'type': s:get_textprop_name(a:server, l:highlight['scope'])})
+        endfor
+    elseif s:use_nvim_highlight
+        " Clear text properties from the previous run
+        call nvim_buf_clear_namespace(a:buf, s:namespace_id, a:line, a:line + 1)
+
+        for l:highlight in a:highlights
+            call nvim_buf_add_highlight(a:buf, s:namespace_id, s:get_hl_name(a:server, l:scopes[l:highlight['scope']]), a:line, l:highlight['char'], l:highlight['char'] + l:highlight['length'])
+        endfor
+    endif
+endfunction
+
+function! s:get_hl_name(server, scope) abort
+    let l:hl = 'LspUnknownScope'
+
+    try
```

Comment:
curious what is the reason for `try..catch` here?

might be good to comment it.

Reply:
The LSP scopes are organised in a tree-like structure:
![image](https://user-images.githubusercontent.com/10748726/71478653-0dc77b00-27f1-11ea-95f6-fe0469c3aef0.png)

So if I want to look up the highlight group for e.g. `entity.name.function.java > meta.function-call.java > ... > source.java` I need to use the most specific setting that the user has defined. The cleanest way is to just iterate in the order `entity.name.function.java`, `entity.name.function.java > meta.function-call.java`, ..., and catch the exception if an entry doesn't exist.

I can also remove the try and add checks for `has_key(a:scope, l:i)` and `has_key(l:info, 'semantic_highlight')`, but I think the try is a little less cluttered.